examples/features/advancedtls/client/main.go

```diff
@@ -0,0 +1,304 @@
+			IdentityOptions: advancedtls.IdentityCertificateOptions{
+				IdentityProvider: identityProvider,
+			},
+			// Setup the roots to be used
+			RootOptions: advancedtls.RootCertificateOptions{
+				RootProvider: rootProvider,
+			},
+			// Tell the client to verify the server cert
+			VerificationType:           advancedtls.CertVerification,
+			AdditionalPeerVerification: customVerificaitonFail,
+		}
+
+		clientTLSCreds, err := advancedtls.NewClientCreds(options)
+
+		if err != nil {
+			fmt.Printf("Error %v\n", err)
+			os.Exit(1)
+		}
+		runWithCredentials(clientTLSCreds, fullServerAddr, false)
+	}
+}
+
+// -- credentials.NewTLS example --
+func credentialsNewTLSExample(credsDirectory string) {
+	cert, err := tls.LoadX509KeyPair(filepath.Join(credsDirectory, "client_cert.pem"), filepath.Join(credsDirectory, "client_key.pem"))
+	if err != nil {
+		os.Exit(1)
+	}
+	rootPem, err := os.ReadFile(filepath.Join(credsDirectory, "ca_cert.pem"))
+	if err != nil {
+		os.Exit(1)
+	}
+	root := x509.NewCertPool()
+	if !root.AppendCertsFromPEM(rootPem) {
+		os.Exit(1)
+	}
+
+	config := &tls.Config{
+		Certificates: []tls.Certificate{cert},
+		RootCAs:      root,
+	}
+
+	// Directly create credentials from a tls.Config.
+	creds := credentials.NewTLS(config)
+	port := goodServerPort
+	fullServerAddr := serverAddr + ":" + port
+	runWithCredentials(creds, fullServerAddr, true)
+
+}
+
+// -- Insecure --
+func insecureCredentialsExample() {
+	creds := insecure.NewCredentials()
+	port := insecurePort
+	fullServerAddr := serverAddr + ":" + port
+	runWithCredentials(creds, fullServerAddr, true)
+}
+
+// -- Main and Runner --
+
+// All of these examples differ in how they configure the
+// credentials.TransportCredentials object. Once we have that, actually making
+// the calls with gRPC is the same.
+func runWithCredentials(creds credentials.TransportCredentials, fullServerAddr string, shouldSucceed bool) {
+	conn, err := grpc.NewClient(fullServerAddr, grpc.WithTransportCredentials(creds))
+	if err != nil {
+		fmt.Printf("Error during grpc.NewClient %v\n", err)
+		os.Exit(1)
+	}
+	defer conn.Close()
+	client := pb.NewEchoClient(conn)
+	req := &pb.EchoRequest{
+		Message: message,
+	}
+	context, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	resp, err := client.UnaryEcho(context, req)
+	defer cancel()
+
+	if shouldSucceed && err != nil {
+		fmt.Printf("Error during client.UnaryEcho %v\n", err)
+	} else if !shouldSucceed && err == nil {
+		fmt.Printf("Should have failed but didn't, got response: %v\n", resp)
+	}
+
+}
+func main() {
+	credsDirectory := flag.String("credentials_directory", "", "Path to the creds directory of this example repo")
+	flag.Parse()
+
+	if *credsDirectory == "" {
+		fmt.Println("Must set credentials_directory argument to this repo's creds directory")
+		os.Exit(1)
+	}
+	tlsWithCrls(*credsDirectory)
+	customVerification(*credsDirectory)
+	credentialsNewTLSExample(*credsDirectory)
+	insecureCredentialsExample()
```

Comment:
Probably too late to change stuff now, but I think it would be good to split things up into multiple clients & servers instead.  This one binary does a lot of different things, and doing so makes the code pretty complicated and hard to understand as an example.  My 2c.

Reply:
Discussed, will leave this for now and potentially make changes in the future